### PR TITLE
X1-AV-40: calc altitude

### PR DESCRIPTION
**Pull request by cjsquires**

### CODE REVIEW

**Comment by cjsquires** [comment]
The altitude calculations mainly function by measuring the lapse rate during the initialization phase, and using it alongside readings of temperature and barometric pressure, alongside the readings of pressure and temperature taken during the initialization phase as the formulas require.

Quick note, lapse rate deals with how the moistness of the air affects how temperature changes with altitude within the troposphere (which we shouldn’t be leaving during flight, lowest boundary of it is 8000 meters).


Definitions of the functions:

initializedLapseRate: uses the values of moistness, pressure, and temperature at the launch pad in order to get a value for lapse rate to use throughout the flight. This should not be changed after launch. 

Had to be initialized as it needs pressure and temperature as inputs, and we need to keep those separate during flight so we can maintain the two different and isolated methods of calculating altitude.


altitudeBarometric: finds altitude from the barometric pressure, using initialized pressure and initialized temperature at launch pad. Should not use any temperature readings from after initialisation phase.


altitudeTemperature_v1: one of two options for calculating altitude from temperature readings. Both use a standard formula that involves change in altitude, change in temperature, and the lapse rate. Both also use riemann sums in order to calculate it.

This version makes a new reimann sum each time it reads, meaning it requires to keep track of its previous readings, which uses a global variable. This variable can only changed in one place, and should never be touched otherwise.

The function also checks for errors in readings, being either null values or the previous temperature reading being lower than the current one. In these cases, the function uses the last change in altitude it calculated to make a rough estimate. Keep in mind, if one temperature value is null, the next run will also fail, as the function uses the current and previous temperature value.

This function should model the temperature change better, but must deal with potentially compounding errors.


altitudeTemperature_v2: same as the previous function, except in how it uses riemann sums. This one makes a singular riemann sum instead of several, treating the initialized altitude and temperature as variables. This one doesn’t have to worry about compounding errors, but it treats the trajectory as linear (at least it should mathematically).


(I really hope this is understandable enough)

File: src/dataHandler.py [hunk 1]
```diff
@@ -60,4 +60,162 @@ def send_all_data(self, sendingTo, destination):
...
+    #previous calculated height change
+    prev_height_change = 0
+    # previous change of height to compensate for Null case
+
+    #lapse rate of the air
+    lapse_rate = 0.0098
+    # this is the standard dry air lapse rate, altered by initializeLapseRate()
+
+
+    def initializeLapseRate(moist, press, temp):
```
Comment (lisaundre) [inline src/dataHandler.py:98]:
```suggestion
    def initialize_lapse_rate(moist, press, temp):
```

snake case function names for consistency!

File: src/dataHandler.py [hunk 2]
```diff
@@ -60,4 +60,162 @@ def send_all_data(self, sendingTo, destination):
...
         
+        lapse_rate = (numerator/denomenator)
+        
+        return lapse_rate
+
+
+
+
+
+    def altitudeBarometric(press, init_press, init_temp): # or altBaro, if needed
```
Comment (lisaundre) [inline src/dataHandler.py:129]:
```suggestion
    def altitude_barometric(press, init_press, init_temp): # or altBaro, if needed
```
snake case again

File: src/dataHandler.py [hunk 3]
```diff
@@ -60,4 +60,162 @@ def send_all_data(self, sendingTo, destination):
...
+
+        elif type(press) == None:
+            alt_baro = None
+
+        return alt_baro
+	
+
+
+
+    def altitudeTemperature_v1(curr_temp, prev_temp, prev_alt): # or altTemp if needed
```
Comment (lisaundre) [inline src/dataHandler.py:156]:
```suggestion
    def altitude_temperature_v1(curr_temp, prev_temp, prev_alt): # or altTemp if needed
```
snake case

File: src/dataHandler.py [hunk 4]
```diff
@@ -60,4 +60,162 @@ def send_all_data(self, sendingTo, destination):
...
+
+            prev_alt = prev_alt + prev_height_change
+
+
+        return alt_temp
+
+
+
+
+    def altitudeTemperature_v2(curr_temp, init_temp): # or altTemp if needed
```
Comment (lisaundre) [inline src/dataHandler.py:202]:
```suggestion
    def altitude_temperature_v2(curr_temp, init_temp): # or altTemp if needed
```
snake case

File: src/dataHandler.py [hunk 5]
```diff
@@ -60,4 +60,162 @@ def send_all_data(self, sendingTo, destination):
...
+        """
+        numerator_numerator = h2o_vapour_heat * moist
+        numerator_denomenator = temp * specif_gas_const_h2o * (press - moist)
+
+        numerator = gravity * (1 + (numerator_numerator/numerator_denomenator))
+
+        denomenator_numerator = h2o_vapour_heat**2 * specif_gas_const_dry * moist
+        denomenator_denomenator = (specif_gas_const_h2o * temp)**2 * (press - moist)
+
+        denomenator = specif_heat_dryair + (denomenator_numerator/denomenator_denomenator)
```
Comment (lisaundre) [inline src/dataHandler.py:119]: i don't know if there's any better way to name these, but it's a little bit confusing to have "numerator_numerator" and "numerator_denominator " and just "numerator"

File: src/dataHandler.py [hunk 6]
```diff
@@ -60,4 +60,162 @@ def send_all_data(self, sendingTo, destination):
...
+
+    #previous calculated height change
+    prev_height_change = 0
+    # previous change of height to compensate for Null case
+
+    #lapse rate of the air
+    lapse_rate = 0.0098
+    # this is the standard dry air lapse rate, altered by initializeLapseRate()
+
+
```
Comment (lisaundre) [inline src/dataHandler.py:97]: can you put the formulas in a docstring?

Comment (lisaundre) [reply src/dataHandler.py:97]: for all of the formulas, not just initializing lapse rate

**Review by lisaundre — COMMENTED** [review]
mostly style consistency/readability things

**Pull request merged**